crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +287,132 @@ impl Report for PrintJson {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct PrintSarif {
+    results: Mutex<RefCell<Vec<sarif::Result>>>,
+    error: Mutex<RefCell<Option<String>>>,
+}
+
+impl Default for PrintSarif {
+    fn default() -> Self {
+        Self {
+            results: Mutex::new(RefCell::new(Vec::new())),
+            error: Mutex::new(RefCell::new(None)),
+        }
+    }
+}
+
+impl Report for PrintSarif {
+    fn report(&self, msg: Message<'_>) -> Result<(), std::io::Error> {
+        match &msg {
+            Message::BinaryFile(_) => {
+                // todo: implement artifacts for sarif
+            }
+            Message::FileType(_) => {
+                // todo: implement artifacts for sarif
+            }
+            Message::File(_) => {
+                // todo: implement artifacts for sarif
+            }
+            Message::Typo(msg) => {
+                let start = String::from_utf8_lossy(&msg.buffer[0..msg.byte_offset]);
+                let column_number =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(start.as_ref(), true).count() + 1;
+                let line_num = msg.context.as_ref().map_or(1, |context| match context {
+                    Context::File(context) => context.line_num,
+                    _ => 1,
+                });
+
+                let message = match &msg.corrections {
+                    typos::Status::Valid => { return Ok(()) }
+                    typos::Status::Invalid => {
+                        format!("`{}` is disallowed", msg.typo)
+                    }
+                    typos::Status::Corrections(corrections) => {
+                        format!("`{}` should be {}", msg.typo, itertools::join(
+                            corrections.iter().map(|s| format!("`{}`", s)),
+                            ", ",
+                        ))
+                    }
+                };
+
+                let location = sarif::LocationBuilder::default()
+                    .physical_location(sarif::PhysicalLocationBuilder::default()
+                        .artifact_location(
+                            sarif::ArtifactLocationBuilder::default()
+                                .uri(msg.context.as_ref().map_or("", |context| match context {
+                                    Context::File(context) => context.path.to_str().unwrap_or(""),
+                                    _ => "",
+                                }))
+                                .build().unwrap()
+                        )
+                        .region(
+                            sarif::RegionBuilder::default()
+                                .start_line(line_num as i64)
+                                .start_column(column_number as i64)
+                                .end_column((column_number + msg.typo.len()) as i64)
```

Comment:
`typo.len()` might not be an appropriate `end_column`.  Since we used `unicode_segmentation::UnicodeSegmentation::graphemes` for start column, we should probably do the same to `msg.typo`.

Granted, all of this depends on what SARIF's definition of a "column" is (bytes, code points, graphemes)

Reply:
Started a separate comment on the units of `column`